soccer/planning/Trajectory.cpp

```diff
@@ -0,0 +1,344 @@
+#include "Trajectory.hpp"
+
+#include <Geometry2d/Pose.hpp>
+#include <stdexcept>
+
+#include "Instant.hpp"
+#include "Utils.hpp"
+
+namespace Planning {
+
+using Geometry2d::Pose;
+using Geometry2d::Twist;
+
+Trajectory::Trajectory(Trajectory a, const Trajectory& b) {
+    if (a.empty() || b.empty()) {
+        throw std::invalid_argument("Cannot splice empty trajectories");
+    }
+
+    RobotInstant a_end = a.last();
+    RobotInstant b_begin = b.first();
+
+    using Geometry2d::Point;
+    if (!a_end.position().nearPoint(b_begin.position(), 1e-6) ||
+        !a_end.linear_velocity().nearPoint(b_begin.linear_velocity(), 1e-6) ||
+        a_end.stamp != b_begin.stamp) {
+        throw std::invalid_argument(
+            "Cannot splice trajectories a and b, where a.last() != b.first()");
+    }
+
+    instants_ = std::move(a.instants_);
+    instants_.reserve(instants_.size() + b.instants_.size() - 1);
+    instants_.insert(instants_.end(), b.instants_.begin() + 1,
+                     b.instants_.end());
+
+    has_angle_profile_ = false;
+    creation_stamp_ = std::nullopt;
+}
+
+void Trajectory::AppendInstant(RobotInstant instant) {
+    if (!empty() && instant.stamp <= end_time()) {
+        throw std::invalid_argument(
+            "Cannot append instant at or before the "
+            "last instant in the trajectory");
+    }
+
+    instants_.push_back(instant);
+
+    // This is not profiled correctly, so we should mark the entire
+    // trajectory's angle profile as invalid.
+    has_angle_profile_ = false;
+    creation_stamp_ = std::nullopt;
+}
+
+bool Trajectory::CheckTime(RJ::Time time) const {
+    return time >= begin_time() && time <= end_time();
+}
+
+bool Trajectory::CheckSeconds(RJ::Seconds seconds) const {
+    return seconds >= 0s && seconds <= duration();
+}
+
+void Trajectory::ScaleDuration(RJ::Seconds final_duration) {
+    // Note: even though this modifies the trajectory, it does not break
+    // angle planning and so we do not need to mark the angle profile as
+    // invalid.
+    ScaleDuration(final_duration, begin_time());
+}
+
+void Trajectory::ScaleDuration(RJ::Seconds final_duration,
```

Comment:
I forgot where this is used, but this will be a bit slower than the old version. It's probably fine though

Reply:
It's not actually used anywhere, I think. Yeah it'll be O(n) instead of O(1) but it's super infrequently used.

Reply:
Yeah that's fine. Trajectories are smaller now. I didn't realize that when making this comment so this is not really an issue.